File: Lesson09_OOP/04_Extension/Clothing.java
/*
Copyright Ann Barcomb and Emily Marasco, 2022
Licensed under GPL v3
See LICENSE.txt for more information.
*/

public class Clothing {

    private int colour;
    private String material;
    private final static String COLOURS[] = {"black", "white", "yellow"};

    // Getters and setters
    public String getColour() { return COLOURS[this.colour]; }
    public int getColourID() { return this.colour; }
    public String getMaterial() { return this.material; }
    public void setMaterial(String material) { this.material = material; }
    public void setColour(String colour) { 
        for (int i=0; i<COLOURS.length; i++) {
            if (colour.equals(COLOURS[i])) {
                this.colour = i;
            } else {
                this.colour = 0;
            }
        }
    }
    public void setColour(int colour) {
        if (colour < COLOURS.length) {
            this.colour = colour;
        } else {
            this.colour = 0;
        }
    }

    // Constructors
    public Clothing(String colour, String material) {
        this.setColour(colour);
        this.material = material;
    }
    public Clothing(int colour, String material) {
        this.setColour(colour);
        this.material = material;
    }

    public static void main(String[] args) {
        var garment = new Clothing(0, "cotton");
        System.out.printf("Colour: %s, Material: %s\n", 
          garment.getColour(), garment.getMaterial());
    }
}

File: Lesson09_OOP/05_Encapsulation/clothing.py
"""
Copyright Khawla Shnaikat and ChatGPT 3.5, 2023
Licensed under GPL v3
See LICENSE.txt for more information.
"""
class Clothing:
    COLOURS = ["black", "white", "yellow"]

    def __init__(self, colour, material):
        if isinstance(colour, int):
            self.set_colour(colour)
        elif isinstance(colour, str):
            self.set_colour(colour)
        self.material = material

    def get_colour(self):
        return self.COLOURS[self.colour]

    def get_colour_id(self):
        return self.colour

    def get_material(self):
        return self.material

    def set_colour(self, colour):
        if isinstance(colour, int):
            self.colour = colour if colour < len(self.COLOURS) else 0
        elif isinstance(colour, str):
            try:
                self.colour = self.COLOURS.index(colour)
            except ValueError:
                self.colour = 0

    def set_material(self, material):
        self.material = material
